### PR TITLE
Fix replicate duplicate metrics

### LINKED ISSUE #2247: bucket replicate panics on prometheus registration
**Thanos, Prometheus and Golang version used**: 
thanosio/thanos:master-2020-03-10-50ffaaaa

**Object Storage Provider**: 
S3

**What happened**:
When attempting to replicate data between two S3 buckets, a panic is reported stating `duplicate metrics collector registration attempted`

**What you expected to happen**:
The process should replicate data without panicking

**How to reproduce it (as minimally and precisely as possible)**:
Ran the following command in a docker environment:
```
bucket replicate --objstore.config-file=/config/from-config.yaml --objstore-to.config-file=/config/to-config.yaml --log.level=debug --single-run
```
Where the config files point to two S3 buckets in the same AWS region
**Full logs to relevant components**:

<details>Logs
<p>

```
level=debug ts=2020-03-10T14:53:52.496252139Z caller=main.go:104 msg="maxprocs: Leaving GOMAXPROCS=[6]: CPU quota undefined"
level=info ts=2020-03-10T14:53:52.496444563Z caller=main.go:153 msg="Tracing will be disabled"
level=debug ts=2020-03-10T14:53:52.496470735Z caller=replicater.go:78 component=replicate msg="setting up http listen-group"
level=info ts=2020-03-10T14:53:52.498942801Z caller=factory.go:46 component=replicate msg="loading bucket configuration"
level=info ts=2020-03-10T14:53:52.501258682Z caller=factory.go:46 component=replicate msg="loading bucket configuration"
level=info ts=2020-03-10T14:53:52.501665456Z caller=replicater.go:208 component=replicate msg="starting replication"
level=info ts=2020-03-10T14:53:52.501793108Z caller=replicater.go:92 component=replicate msg="Listening for http service" address=0.0.0.0:10902
level=info ts=2020-03-10T14:53:52.501966695Z caller=intrumentation.go:60 component=replicate msg="changing probe status" status=healthy
level=info ts=2020-03-10T14:53:52.502053157Z caller=http.go:56 component=replicate service=http/server component=replicate msg="listening for requests and metrics" address=0.0.0.0:10902
level=info ts=2020-03-10T14:53:52.502134293Z caller=replicater.go:171 component=replicate replication-run-id=01E32F04FNKG36J2NXH571FJ3D msg="running replication attempt"
level=debug ts=2020-03-10T14:53:52.502186639Z caller=scheme.go:173 component=replicate replication-run-id=01E32F04FNKG36J2NXH571FJ3D msg="scanning blocks available blocks for replication"
level=debug ts=2020-03-10T14:53:53.427112952Z caller=fetcher.go:358 component=replicate replication-run-id=01E32F04FNKG36J2NXH571FJ3D component=block.MetaFetcher msg="successfully fetched block metadata" duration=525.196648ms cached=14 returned=14 partial=0
level=debug ts=2020-03-10T14:53:53.427249937Z caller=scheme.go:204 component=replicate replication-run-id=01E32F04FNKG36J2NXH571FJ3D msg="adding block to available blocks" block_uuid=01E1TMA6ABM79ENM26R8NQSJ6A
panic: duplicate metrics collector registration attempted

goroutine 131 [running]:
github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc0001fea00, 0xc000a38430, 0x1, 0x1)
	/go/pkg/mod/github.com/prometheus/client_golang@v1.5.0/prometheus/registry.go:400 +0xad
github.com/prometheus/client_golang/prometheus/promauto.Factory.NewCounter(0x22c42e0, 0xc0001fea00, 0x0, 0x0, 0x1ea614a, 0xb, 0x1ea6bf2, 0xb, 0x1ef4f43, 0x2e, ...)
	/go/pkg/mod/github.com/prometheus/client_golang@v1.5.0/prometheus/promauto/auto.go:263 +0x12f
github.com/thanos-io/thanos/pkg/block.newSyncMetrics(0x22c42e0, 0xc0001fea00, 0xc00096f4a0)
	/go/src/github.com/thanos-io/thanos/pkg/block/fetcher.go:61 +0xa9
github.com/thanos-io/thanos/pkg/block.NewMetaFetcher(0x228a7e0, 0xc000ab2030, 0x20, 0x7f33dad9d0a0, 0xc000a791d0, 0x0, 0x0, 0x22c42e0, 0xc0001fea00, 0x0, ...)
	/go/src/github.com/thanos-io/thanos/pkg/block/fetcher.go:139 +0x10e
github.com/thanos-io/thanos/pkg/replicate.loadMeta(0x22d95a0, 0xc000156680, 0xc0000d80a0, 0x37d590d22da97001, 0x177a7b40c0b9de22, 0xc000138390, 0x0, 0x1, 0x0)
	/go/src/github.com/thanos-io/thanos/pkg/replicate/scheme.go:353 +0x8f
github.com/thanos-io/thanos/pkg/replicate.(*replicationScheme).execute.func1(0xc000c32140, 0x1b, 0x1, 0x0)
	/go/src/github.com/thanos-io/thanos/pkg/replicate/scheme.go:185 +0xfd
github.com/thanos-io/thanos/pkg/objstore/s3.(*Bucket).Iter(0xc0001fecd0, 0x22d95a0, 0xc000156680, 0x0, 0x0, 0xc000ab2180, 0x30, 0x1ce0760)
	/go/src/github.com/thanos-io/thanos/pkg/objstore/s3/s3.go:243 +0x2c0
github.com/thanos-io/thanos/pkg/objstore.(*metricBucket).Iter(0xc000a791d0, 0x22d95a0, 0xc000156680, 0x0, 0x0, 0xc000ab2180, 0xc000ab2120, 0x8)
	/go/src/github.com/thanos-io/thanos/pkg/objstore/objstore.go:250 +0x6f
github.com/thanos-io/thanos/pkg/replicate.(*replicationScheme).execute(0xc0000d80a0, 0x22d95a0, 0xc000156680, 0x7f33dad9d0a0, 0xc000a791d0)
	/go/src/github.com/thanos-io/thanos/pkg/replicate/scheme.go:175 +0x1f4
github.com/thanos-io/thanos/pkg/replicate.RunReplicate.func3(0xc00038de38, 0x2307da0)
	/go/src/github.com/thanos-io/thanos/pkg/replicate/replicater.go:173 +0x609
github.com/thanos-io/thanos/pkg/replicate.RunReplicate.func4(0x0, 0x0)
	/go/src/github.com/thanos-io/thanos/pkg/replicate/replicater.go:185 +0x23c
github.com/oklog/run.(*Group).Run.func1(0xc00097c1e0, 0xc00003b730, 0xc000a7d460)
	/go/pkg/mod/github.com/oklog/run@v1.0.0/group.go:38 +0x27
created by github.com/oklog/run.(*Group).Run
	/go/pkg/mod/github.com/oklog/run@v1.0.0/group.go:37 +0xbb
```

</p>
</details>

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Fixes #2247

### CODE REVIEW

File: pkg/replicate/replicater.go [hunk 1]
```diff
@@ -149,6 +150,11 @@ func RunReplicate(
 		Help: "The Duration of replication runs split by success and error.",
 	}, []string{"result"})
 
+	fetcher, err := thanosblock.NewMetaFetcher(logger, 32, fromBkt, "", reg)
```
Comment (bwplotka) [inline pkg/replicate/replicater.go]: Also would be nice to rename the file to `replicator.go` :thinking:

File: pkg/replicate/replicater.go [hunk 2]
```diff
@@ -149,6 +150,11 @@ func RunReplicate(
...
+	if err != nil {
+		return errors.Errorf("create meta fetcher with bucket %v: %w", fromBkt, err)
```
Comment (bwplotka) [inline pkg/replicate/replicater.go]: errors.Wrap is our way to wrap errros (:

File: pkg/replicate/scheme.go [hunk 3]
```diff
@@ -350,12 +360,7 @@ func (rs *replicationScheme) ensureObjectReplicated(ctx context.Context, objectN
 // partial, this is just a temporary failure, as the block is still being
 // uploaded to the origin bucket.
 func loadMeta(ctx context.Context, rs *replicationScheme, id ulid.ULID) (*metadata.Meta, bool, error) {
-	fetcher, err := thanosblock.NewMetaFetcher(rs.logger, 32, rs.fromBkt, "", rs.reg)
```
Comment (bwplotka) [inline pkg/replicate/scheme.go]: Wow this was wrong :scream:

Comment (bwplotka) [reply pkg/replicate/scheme.go]: How it missed our review...

File: pkg/replicate/scheme_test.go [hunk 4]
```diff
@@ -314,11 +315,16 @@ func TestReplicationSchemeAll(t *testing.T) {
 		}
 
 		filter := NewBlockFilter(logger, selector, compact.ResolutionLevelRaw, 1).Filter
+		fetcher, err := block.NewMetaFetcher(logger, 32, originBucket, "", nil)
+		if err != nil {
```
Comment (bwplotka) [inline pkg/replicate/scheme_test.go]: `testutil.Ok`?

File: pkg/replicate/replicater.go (hunk 2 above)
Comment (GiedriusS) [reply pkg/replicate/replicater.go]: :+1: and this will become an error on `go vet` with 1.14

**Comment by yeya24** [comment]
Hi, @bwplotka  @GiedriusS 
I have addressed your comments. It would be great if you have time to review it again 😄

**Pull request merged**